### ISSUE TITLE
Dirty not being set on first click (component using an object)

### ISSUE THREAD

**Issue by patrickdavey** [monterail/vuelidate#182]
I suspect I'm misunderstanding something basic, but, I have a `component` I am passing a `prop` into. Inside that component (in the `data` method) I take a copy of that object, so that I can mutate it and test it. This object might have lots of properties, so, I created a `computed` property so I can `watch` it and then update my `vuex` store if the form is valid (and perform API requests etc.)

The issue I'm having [which you can see in this jsfiddle](https://jsfiddle.net/patrickjsfiddle/xx7s5bdd/) is that I am attempting to `touch` a checkbox on the `@change` event.

```javascript
Vue.component('test', {
  props: ['obj'],
  data: function () {
    return { form: {
      foo: false
    }};
  },
  computed: {
    formValues () {
      return _.values(this.form);
    }
  },
  watch: {
    formValues: function (value, _oldValue) {
      console.log({
        dirty: this.$v.$dirty
      });
      this.$v.$reset();
      console.log({
        dirty: this.$v.$dirty
      });
    }
  },
 validations: {
  	var: {
      minLength: minLength(5)
    }
  },
  template: `<div><input @change="$v.$touch()" v-model="form.foo" type="checkbox">
{{ form.foo }}</div>`
})

```

The issue I'm seeing is that in the `watch` above, the first time it it called (after the checkbox is clicked)  `$dirty is false`. _for every subsequent checkbox click, `$dirty is true` (then gets reset).

I'm sure I'm doing something evil here, but, any pointers would be most appreciated. Basically I want to re-use this form and when the prop changes, I want to take a fresh clone of the object passed in, and reset the form. Then, I want to watch for any changes to any of the fields on the form and then _automatically_ update the API. If I had a submit button (not what the client wants) this would all be a lot easier.

Thank you very much for any help you can give, and for such an excellent plugin.

**Comment by patrickdavey** [comment]
As a quick addendum, if I change [to this later version](https://jsfiddle.net/patrickjsfiddle/xx7s5bdd/7/) then I do get the dirty flag set correctly from the start.

Again, it's probably just me, but I do find it confusing!

#### changes to make it work

```javascript
  methods: {
    updateValue (property) {
      this.$v.$touch();
      let updatedCopy = Object.assign({}, this.form, property);
      this.form = updatedCopy;
    }
  },

//and

 template: `<div> <input
                     v-bind:value="form.foo"
                     type="checkbox"
                     v-on:change="updateValue({foo: $event.target.value})">`
```

**Comment by Frizi** [comment]
Well, it all boils down to how you use your browser events. The v-model on input field is equivalent to setting the value on `@input` event. Your "watch" triggers on value change, which is simply a next microtask ahead of that in the event loop (`nextTick` from vue). That still is before the `@change` happens. There is nothing wrong with vuelidate. Treat the `$touch` as setting `dirty` to true, as that's what it literally is doing.

https://github.com/monterail/vuelidate/blob/c60e948076d4852b9a07f53a58fbd1adf2f36a20/src/index.js#L88-L103

Your second example kinda works because you change your "sampling time" of those values.

I would advise against that practice altogether, and instead focus on what you want to accomplish in general. What's your usecase? There is certainly a better and safer way to do it.

**Comment by patrickdavey** [comment]
Many thanks for taking the time to respond, especially when this is just me using vue badly! As you asked about my use case, I'll answer, but please feel free to close this and I can just ask in the forum - I don't want to waste your time!

My usecase is autosaving a form. Basically I have a list of items. When an item is clicked on I display a form for the corresponding item (the `prop` which is passed in I `watch` and then re-initialize a `data: form` object). It is desired to have this form "auto save" (i.e. no explicit save button). My watch actually looks like this (problem is the prop passed in)

```javascript
  watch: {
    problem (problemObject, _oldValue) {
      this.form = Object.assign({}, _.pick(problemObject, "override_objective", "override_fullwidth",
        "start_on_new_page", "hide_choices", "workspace_value"));
      this.$v.$reset();
    },

    formValues (value, _oldValue) {
      if (!this.$v.$error && this.$v.$dirty) {
        this.$v.$reset();
        this.$store.dispatch("UPDATE_PART_PROBLEM_FORMAT", {
          partId: this.problem.id,
          properties: this.form
        });
      }
    }
  }

```

So, basically when the `formValues` object changes, then, I want to check if the form is valid & dirty, and if so then dispatch an action. The issue I was having is that because `dirty` isn't set (because of the timing issue above) then the store doesn't get updated.

Many thanks for taking the time to reply!

**Comment by Frizi** [comment]
You should probably be doing a `debounce` on that form anyway. My approach on that would be to:
- somehow modify your data as a response to the form (probably each field could have associated lens/two methods, which would tell it how to read and write data to/from associated store)
- issue debounced save action alongside every triggered store dispatch action
- in the save handler, first check for `this.$v.$invalid` and only then send the bulk update request.

Other way would also be to straight copy the data from store, then modify the data using v-model, and issue the debounced save in the deep watch. Depends on your architecture.
